Rename topDocument() and make it return a pointer instead of a reference

beidson opened the pull request:
#### 18391053bff1f5c0ebc6597d9fc3e693e6168cdd
<pre>
Rename topDocument() and make it return a pointer instead of a reference
<a href="https://rdar.apple.com/142888352">rdar://142888352</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=285917">https://bugs.webkit.org/show_bug.cgi?id=285917</a>

Reviewed by Alex Christensen.

This accomplishes two things:
1 - Makes searching for true &quot;topDocument()&quot; uses in the codebase much easier since the phrase &quot;mainFrameDocument&quot; doesn&apos;t otherwise exist.
2 - Instead of always returning a Document&amp; even if it&apos;s actually wrong, this forces us to confront the reality that a mainFrameDocument
    often doesn&apos;t exist with site isolation enabled, so we have to handle each case reasonably.

* Source/WTF/wtf/Assertions.h:
* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(WebCore::PaymentSession::canCreateSession):
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::shouldDocumentAllowWebAudioToAutoPlay):
(WebCore::AudioContext::constructCommon):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::topDocument const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityContainer]):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::~Document):
(WebCore::Document::topOrigin const):
(WebCore::Document::createRenderTree):
(WebCore::Document::destroyRenderTree):
(WebCore::Document::willBeRemovedFromFrame):
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
(WebCore::Document::implicitClose):
(WebCore::Document::urlForBindings):
(WebCore::Document::flushAutofocusCandidates):
(WebCore::Document::referrerForBindings):
(WebCore::Document::mainFrameDocument const):
(WebCore::Document::shouldForceNoOpenerBasedOnCOOP const):
(WebCore::Document::mainFrameDocumentHasHadUserInteraction const):
(WebCore::Document::processingUserGestureForMedia const):
(WebCore::Document::deviceOrientationAndMotionAccessController):
(WebCore::Document::isRunningUserScripts const):
(WebCore::Document::setAsRunningUserScripts):
(WebCore::Document::crossOriginOpenerPolicy const):
(WebCore::Document::advancedPrivacyProtections const):
(WebCore::Document::topDocument const): Deleted.
* Source/WebCore/dom/Document.h:
(WebCore::Document::protectedMainFrameDocument const):
(WebCore::Document::isTopDocumentLegacy const):
(WebCore::Document::protectedTopDocument const): Deleted.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::insertedIntoAncestor):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::mainFrameDocument):
(WebCore::FullscreenManager::protectedMainFrameDocument):
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
(WebCore::FullscreenManager::protectedTopDocument): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/dom/UserGestureIndicator.cpp:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_remote):
(WebCore::HTMLMediaElement::removeBehaviorRestrictionsAfterFirstUserGesture):
(WebCore::HTMLMediaElement::updateRateChangeRestrictions):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::isReplacementObscured):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::playbackStateChangePermitted const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::keyEvent):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::crossOriginIsolated const):
(WebCore::LocalDOMWindow::incrementScrollEventListenersCount):
(WebCore::LocalDOMWindow::decrementScrollEventListenersCount):
* Source/WebCore/page/Quirks.cpp:
(WebCore::allowedAutoplayQuirks):
(WebCore::Quirks::needsAutoplayPlayPauseEvents const):
(WebCore::Quirks::needsPerDocumentAutoplayBehavior const):
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::isHiddenBehindFullscreenElement):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::RepaintRegionAccumulator::RepaintRegionAccumulator):
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::policySourceDocumentLoader const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::isObscuredElement):

Canonical link: <a href="https://commits.webkit.org/288984@main">https://commits.webkit.org/288984@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb4d7597b15577a661700f31c5203be4f4befb59

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84940 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39330 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90089 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35997 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87025 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12642 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/23922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3637 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77196 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3511 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31423 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35069 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77914 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32231 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91468 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83983 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12279 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12509 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73005 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73712 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18244 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18094 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16544 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12231 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106375 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12066 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25677 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15569 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13817 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5700,8 +5714,11 @@ void Document::flushAutofocusCandidates()
+        if (!elementMainFrameDocument)
+            LOG_ERROR_ONCE("Unable to fully perform Document::flushAutofocusCandidates() without access to the elements main frame document (site isolation)");
+
+        if (!element || !element->document().isFullyActive() || (elementMainFrameDocument && elementMainFrameDocument != this)) { m_autofocusCandidates.removeFirst();
```
m_autofocusCandidates.removeFirst() should be on next line.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10633,17 +10669,27 @@ DOMTimerHoldingTank& Document::domTimerHoldingTank()
+        mainFrameDocument->setAsRunningUserScripts();
     else
-        top->setAsRunningUserScripts();
+        LOG_ERROR_ONCE("Unable to properly set Document::setAsRunningUserScripts() without access to the main frame document (site isolation)");
```
Do any of these print when running SiteIsolation API tests?

beidson replied:
They all do (but, just once per process)

The last time I caused issues with the tests with logging, it was unbounded logging - This was an attempt to resolve that?

achristensen07 commented on Source/WTF/wtf/Assertions.h:
```diff
@@ -585,19 +585,37 @@ static constexpr bool unreachableForValue = false;
 
 #if ERROR_DISABLED
 #define LOG_ERROR(...) ((void)0)
+#define LOG_ONCE(SiteIsolation, ...) ((void)0)
```
channel, not SiteIsolation.  Or it could just be ...

beidson replied:
Whoops, that one fixed locally dunno how the upload missed it.

achristensen07 commented on Source/WebCore/Modules/applepay/PaymentSession.cpp:
```diff
@@ -58,9 +58,14 @@ ExceptionOr<void> PaymentSession::canCreateSession(Document& document)
+        return Exception { ExceptionCode::InvalidAccessError, "Trying to start an Apple Pay session from a site isolated iframe"_s };
+    }
+
+    if (&document != mainFrameDocument.get()) {
```
isTopDocument?

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4191,7 +4199,13 @@ const URL& Document::urlForBindings()
-        RefPtr policySourceLoader = topDocument().loader();
+        RefPtr mainFrameDocument = protectedMainFrameDocument();
+        if (!mainFrameDocument) {
+            LOG_ONCE(SiteIsolation, "Unable to completely calculate Document::urlForBindings() without access to the main frame document (site isolation)");
```
(site isolation) is now probably redundant.  Same throughout.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7385,21 +7409,21 @@ Document* Document::parentDocument() const
         if (RefPtr localMainFrame = this->localMainFrame())
             localMainDocument = localMainFrame->document();
-        return localMainDocument ? *localMainDocument : const_cast<Document&>(*this);
+        return localMainDocument ? localMainDocument : const_cast<Document*>(this);
```
We may eventually want a "const Document* mainFrameDocument()" const and "Document* mainFrameDocument()" but definitely not in this PR.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7832,8 +7856,11 @@ bool Document::shouldForceNoOpenerBasedOnCOOP() const
 
-    auto COOPValue = topDocument().crossOriginOpenerPolicy().value;
-    return (COOPValue == CrossOriginOpenerPolicyValue::SameOrigin || COOPValue == CrossOriginOpenerPolicyValue::SameOriginPlusCOEP) && !isSameOriginAsTopDocument();
+    CrossOriginOpenerPolicyValue coopValue = { };
```
auto coopValue = CrossOriginOpenerPolicyValue::UnsafeNone

achristensen07 commented on Source/WebCore/html/HTMLMediaElement.h:
```diff
@@ -1288,8 +1288,8 @@ class HTMLMediaElement
     bool m_processingPreferenceChange : 1;
-    bool m_shouldAudioPlaybackRequireUserGesture : 1;
-    bool m_shouldVideoPlaybackRequireUserGesture : 1;
+    bool m_shouldAudioPlaybackRequireUserGesture : 1 { true };
```
Initializing bit fields with initializer lists?  Has technology really come this far?  Back in my day this was impossible.

beidson replied:
👍

beidson committed:
Rename topDocument() and make it return a pointer instead of a reference
rdar://142888352
https://bugs.webkit.org/show_bug.cgi?id=285917

Reviewed by Alex Christensen.

This accomplishes two things:
1 - Makes searching for true "topDocument()" uses in the codebase much easier since the phrase "mainFrameDocument" doesn't otherwise exist.
2 - Instead of always returning a Document& even if it's actually wrong, this forces us to confront the reality that a mainFrameDocument
    often doesn't exist with site isolation enabled, so we have to handle each case reasonably.

* Source/WTF/wtf/Assertions.h:
* Source/WebCore/Modules/applepay/PaymentSession.cpp:
(WebCore::PaymentSession::canCreateSession):
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::shouldDocumentAllowWebAudioToAutoPlay):
(WebCore::AudioContext::constructCommon):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::topDocument const):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityContainer]):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::~Document):
(WebCore::Document::topOrigin const):
(WebCore::Document::createRenderTree):
(WebCore::Document::destroyRenderTree):
(WebCore::Document::willBeRemovedFromFrame):
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
(WebCore::Document::implicitClose):
(WebCore::Document::urlForBindings):
(WebCore::Document::flushAutofocusCandidates):
(WebCore::Document::referrerForBindings):
(WebCore::Document::mainFrameDocument const):
(WebCore::Document::shouldForceNoOpenerBasedOnCOOP const):
(WebCore::Document::mainFrameDocumentHasHadUserInteraction const):
(WebCore::Document::processingUserGestureForMedia const):
(WebCore::Document::deviceOrientationAndMotionAccessController):
(WebCore::Document::isRunningUserScripts const):
(WebCore::Document::setAsRunningUserScripts):
(WebCore::Document::crossOriginOpenerPolicy const):
(WebCore::Document::advancedPrivacyProtections const):
(WebCore::Document::topDocument const): Deleted.
* Source/WebCore/dom/Document.h:
(WebCore::Document::protectedMainFrameDocument const):
(WebCore::Document::isTopDocumentLegacy const):
(WebCore::Document::protectedTopDocument const): Deleted.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::insertedIntoAncestor):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::mainFrameDocument):
(WebCore::FullscreenManager::protectedMainFrameDocument):
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::cancelFullscreen):
(WebCore::FullscreenManager::exitFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
(WebCore::FullscreenManager::protectedTopDocument): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/dom/UserGestureIndicator.cpp:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::runFocusingSteps):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::runPopoverFocusingSteps):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::m_remote):
(WebCore::HTMLMediaElement::removeBehaviorRestrictionsAfterFirstUserGesture):
(WebCore::HTMLMediaElement::updateRateChangeRestrictions):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::isReplacementObscured):
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::playbackStateChangePermitted const):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::keyEvent):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::crossOriginIsolated const):
(WebCore::LocalDOMWindow::incrementScrollEventListenersCount):
(WebCore::LocalDOMWindow::decrementScrollEventListenersCount):
* Source/WebCore/page/Quirks.cpp:
(WebCore::allowedAutoplayQuirks):
(WebCore::Quirks::needsAutoplayPlayPauseEvents const):
(WebCore::Quirks::needsPerDocumentAutoplayBehavior const):
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::isHiddenBehindFullscreenElement):
* Source/WebCore/platform/Logging.h:
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::RepaintRegionAccumulator::RepaintRegionAccumulator):
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::policySourceDocumentLoader const):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::isObscuredElement):

Canonical link: https://commits.webkit.org/288984@main
